doc/architectural/background-concepts.md

```diff
@@ -764,16 +764,378 @@ Next, try running minisat on the following formula:
+propositional formula does not actually describe the order of execution of the
+statements, but simply summarizes their outcomes! Once we have non-trivial
+control flow, we have to do some extra work in this model.
+
+One example of non-trivial control flow are `if` statements. Consider the
+following example:
+```C
+int max(int a, int b)
+{
+  int result;
+  if (a < b)
+    result = b;
+  else
+    result = a;
+  return result;
+}
+```
+Bringing this into SSA form, we have the following program (we write `Phi` for &Phi;):
+```C
+int max(int a, int b)
+{
+  int result;
+  if (a < b)
+    result.1 = b;
+  else
+    result.2 = a;
+  return Phi(result.1,result.2);
+}
+```
+We again introduce bit vectors A (for parameter `a`), B (for parameters `b`),
+R1 (for `result.1`), R2 (for `result.2`) and
+return (for the return value). The interesting question in this case is
+how we can handle the &Phi; node: so far, it is a ''magic'' operator that selects
+the correct value.
+
+As a first step, we modify the SSA form slightly by introducing an additional
+propositional variable C that tracks which branch of the `if` was taken.
+We call this variabel the *code guard variable*, or *guard* for short.
+Additionally, we add C to the &Phi; node as a new first parameter, describing
+which input to use as a result.
+The corresponding program looks something like this:
+```C
+int max(int a, int b)
+{
+  int result;
+  bit C; /* Track which branch was taken */
+  C = a < b;
+  if (C)
+    result.1 = b;
+  else
+    result.2 = a;
+  return Phi(C,result.1,result.2);
+}
+```
+For the encoding of the program, we introduce a new propositional junctor,
+&rArr;, where ''A &rArr; B'' is equivalent to ''(not A) or B''.
+It can be understood as ''if A holds, B must hold as well''.
+
+With these ingredients, we can encode the program. First of all, we translate
+the basic statements of the program:
+- `C = a<b` maps to C = A&lt;B, for an appropriate formula A&lt;B.
+- `result.1 = b` becomes R1 = B, and `result.2 = a` becomes R2 = A.
+
+To handle the `if` statement, we simply make the execution of each branch
+conditional on C using the &rArr; junctor:
+```C
+  if (C)
+    result.1 = b;
+  else
+    result.2 = a;
+```
+becomes (C &rArr; R1 = B) and ((not C) &rArr; R2 = A), stating that
+the equation for the first assignment holds when C is true, and that for
+the second assignment holds when C is false.
+
+Finally, the &Phi; node is again resolved using the &rArr; junctor: we
+can encode the `return` statement as
+(C &rArr; return = R1) and ((not C) &rArr; return = R2).
+
+At this point, it remains to tie the statements together; we find that we can
+again simply connect them with ''and'', since the statements are always executed
+in sequence. We get:
+> C = a&lt;b and (C &rArr; R1 = B) and (C &rArr; return = R1) and
+> ((not C) &rArr; R2 = A) and ((not C) &rArr; return = R2).
+
+We can extend this approach quite straightforwardly to other constructs, but
```

Comment:
Before going to loops, you will need to introduce properties/assertions. Otherwise you have program constraints with nothing to prove about them. Or maybe actually Hoare Logic with pre- and post-conditions, which actually is the reasoning that's happening.

Reply:
I will introduce ASSERT as a primitive; this also helps me deal with the description of loop unrolling.